### ISSUE TITLE
Deprecation logs in the monitoring UI

### ISSUE THREAD

**Issue by chrisronline** [elastic/stack-monitoring#17]
We want to show deprecation logs in the monitoring UI but there are a few issues we need to discuss and find solutions for.

As a basic premise, the idea is that filebeat will collect deprecation logs and index them into ES. Then, the monitoring UI will simply query for the data and show a link in the UI if there are deprecation logs that will take the user to the logs UI.

The problem with this is in a [recommended monitoring setup](https://www.elastic.co/guide/en/x-pack/current/monitoring-production.html), there is at least one production cluster and one dedicated monitoring cluster. Filebeat [only supports a single output](https://www.elastic.co/guide/en/beats/filebeat/current/configuring-output.html) so it's possible the Filebeat output is a different cluster than the one the monitoring UI reads from (which is the monitoring cluster). In this scenario, there may be deprecation logs but none would be detected and therefore nothing in the UI about it.

To list out the problems, they are:

1) From the monitoring UI, we do not know for certain which ES cluster a filebeat deprecation log document originated from.

2) If the filebeat documents live on a different cluster from the monitoring one, we need a way to read those.

3) (Separate from the above conversation) The logs UI needs to support some sort of deep linking so the link will only show the relevant deprecation messages. In addition to needing a deep linking, we also need to filter off `cluster_uuid` because of problem 1. See https://github.com/elastic/kibana/issues/23316

We have a couple solutions in mind:

1) Live with the restriction and document that users will need to configure their filebeat (or additional filebeat) to output to the _monitoring_ cluster so the monitoring UI can read it.
    - A downside here is that we're making an assumption (maybe small) that filebeat is reading from _and_ indexing into the same cluster, which is not always the case.
    - Another downside is that users might need to adjust their configs to support this, or possibly spin up a new filebeat to output to the monitoring cluster (assuming they need filebeat also indexing into the production cluster)

2) Push [this initiative forward](https://github.com/elastic/elasticsearch/issues/32850#issuecomment-447891359) and update Filebeat to collect and index the `cluster_uuid`. Once this is setup, we'll know definitively which cluster the filebeat read from (solves problem 1). Then, we instruct users to configure cross-cluster search _from_ the monitoring cluster to all production clusters so we can perform a cross-cluster search within the monitoring UI code to discover all filebeat indices.

Personally, I feel the second solution is viable and should be our goal. There are [some limitations](https://github.com/elastic/elasticsearch/issues/32850#issuecomment-447932949) that I don't think are worth switching away from solution 2, but worth discussing if necessary.

We're open to thoughts/comments/suggestions and would love to hear if there are other possible solutions.

cc @elastic/stack-monitoring

**Comment by ppf2** [comment]
Deprecation logs can generate a ton of duplicated entries depending on how many API calls they are making that are deprecated.  If we show it on the UI, it will be great if we can provide a quick way to dedup them so admins can focus on the unique entries :)

**Comment by ycombinator** [comment]
I don't think solution 1 and 2 are either-or. Even with solution 1, we'll still want each log entry document to have a `cluster_uuid` field. Otherwise the Monitoring UI won't be able to distinguish between deprecation log entries from one ES cluster vs. another, in a scenario where a user has multiple "production" ES clusters.

So I see solution 2 as more about using CCS, not about collecting the `cluster_uuid`. I would extract the `cluster_uuid` collection bit out and make it a prerequisite of either solution 1 or 2.

**Comment by cachedout** [comment]
Bear in mind that my overall understanding remains comically small but I'd like to talk through problem 2 in your list.

To restate this, it's possible that a Beat may be configured to a output to an ES cluster. Let's call these `B1` and `ES1`. As you mention, we have a monitoring cluster which we'll call `M1`.

The problem you're describing as I understand it is that B1 may emit events to ES1 and _some_ of those events may be such that we wish them to be seen on M1.

The only relationship between ES1 and M1 is that (in our intended future default topology) is that ES1 sends monitoring data to M1 via Metricbeat configured with the Elasticsearch plugin.

So, my question is this -- what options do we have for ES1 to *know* that it has received an event which M1 should be aware of? Could we use a Watcher to do this?

**Comment by chrisronline** [comment]
@cachedout 

> The problem you're describing as I understand it is that B1 may emit events to ES1 and some of those events may be such that we wish them to be seen on M1.

I wouldn't say this is true. In your scenario, B1 may output events to ES1, but the event B1 is outputting is a file read from a deprecation log on M1. In this case, ES1 _knows_ about the document and that it is a deprecation log, but we need to know that B1 read it from M1 instead of incorrectly assuming B1 read it from ES1.

**Comment by chrisronline** [comment]
@ycombinator 

Yes you're right. There is no real scenario where we're okay not knowing that piece of information. I'm going to update this later tonight

**Comment by chrisronline** [comment]
## Update

We need to fix a core problem with either solution, and honestly, both solutions are somewhat the same in reality.

We need to have the `cluster_uuid` present in the deprecation log. See this issue: https://github.com/elastic/elasticsearch/issues/32850#issuecomment-447891359

Once we have this, we can move forward with a CCS-based solution to discover all filebeat indices for all clusters from the monitoring UI.

**Comment by chrisronline** [comment]
Here are some related PRs to track around this effort:

https://github.com/elastic/beats/pull/9260
https://github.com/elastic/beats/pull/9163

**Comment by ycombinator** [comment]
https://github.com/elastic/elasticsearch/issues/32850 refers to PR https://github.com/elastic/elasticsearch/pull/36833, which is about **structured** ES logs. As Przemyslaw noted those logs will have the cluster UUID in them. And you pointed out that doesn't help us display deprecation logs in the Monitoring UI for 6.7.

The only way forward I see for properly displaying deprecation logs in the Monitoring UI for 6.7 is if ES could add the cluster UUID in their current (i.e. 6.x) **unstructured** deprecations logs as well. Then we can adjust the filebeat elasticsearch module for 6.7 as well to parse+ingest said cluster UUID, and ultimately properly display them in/from the Monitoring UI. So I'd suggest creating a separate ES issue about this bit.

**Comment by cachedout** [comment]
Moving to 6.7 for now because this clearly isn't happening for 6.6 :) We'll just have to wait to hear back from the ES folks.
